extensions/amp-bind/0.1/bind-impl.js

```diff
@@ -119,50 +111,93 @@ export class Bind {
+   * @private
+   */
+  initialize_() {
+    this.scanPromise_ = this.scanForBindings_(this.ampdoc.getBody());
+    this.scanPromise_.then(results => {
+      const {boundElements, evaluatees} = results;
+
+      this.boundElements_ = boundElements;
+      this.evaluator_ = new BindEvaluator(evaluatees);
+
+      // Trigger verify-only digest in development.
+      const development = getMode().development;
+      if (development || this.digestQueuedAfterScan_) {
+        this.digest_(/* opt_verifyOnly */ development);
+      }
+    });
+  }
+
+  /**
    * Scans `body` for attributes that conform to bind syntax and returns
    * a tuple containing bound elements and binding data for the evaluator.
    * @param {!Element} body
-   * @return {{
-   *   boundElements: !Array<BoundElementDef>,
-   *   evaluatees: !Array<./bind-evaluator.EvaluateeDef>,
-   * }}
+   * @return {
+   *   !Promise<{
+   *     boundElements: !Array<BoundElementDef>,
+   *     evaluatees: !Array<./bind-evaluator.EvaluateeDef>
+   *   }>
+   * }
    * @private
    */
   scanForBindings_(body) {
-    // TODO(choumx): Chunk if taking too long in a single frame.
-
     /** @type {!Array<BoundElementDef>} */
     const boundElements = [];
     /** @type {!Array<./bind-evaluator.EvaluateeDef>} */
     const evaluatees = [];
 
+    // TODO(choumx): Use TreeWalker if this is too slow.
     const elements = body.getElementsByTagName('*');
-    for (let i = 0; i < elements.length; i++) {
-      const element = elements[i];
-      const attributes = element.attributes;
-
-      const bindings = [];
-      for (let j = 0; j < attributes.length; j++) {
-        const binding = this.bindingForAttribute_(attributes[j], element);
-        if (binding) {
-          bindings.push(binding);
-          evaluatees.push({
-            tagName: element.tagName,
-            property: binding.property,
-            expressionString: binding.expressionString,
-          });
+
+    // Helper function for scanning a slice of elements.
+    const scanFromTo = (start, end) => {
+      for (let i = start; i < Math.min(end, elements.length); i++) {
+        const el = elements[i];
+        const attrs = el.attributes;
+
+        const bindings = [];
+        for (let j = 0; j < attrs.length; j++) {
+          const binding = this.bindingForAttribute_(attrs[j], el);
+          if (binding) {
+            bindings.push(binding);
+            evaluatees.push({
+              tagName: el.tagName,
+              property: binding.property,
+              expressionString: binding.expressionString,
+            });
+          }
+        }
+        if (bindings.length > 0) {
+          boundElements.push({element: el, bindings});
         }
       }
-      if (bindings.length > 0) {
-        boundElements.push({element, bindings});
-      }
+    };
+
+    // Divide elements into buckets to be scanned, one bucket per chunk.
+    const bucketSize = 10;
```

Comment:
I think this is not the right way to do this when you design for `requestIdleCallback`, but this is a perfect use case for RIC!

The current chunking code has the issue that all tasks that we scheduled could not be split up further (easily). Here you can basically do work until you shouldn't anymore. With that the `timeRemaining` function should be made available to the scanner and it should always try to do as much as it can given the time available.